docker/gcp_pudl_etl.sh

```diff
@@ -79,23 +79,54 @@ function upload_to_dist_path() {
     GCS_PATH="gs://pudl.catalyst.coop/$1/"
     AWS_PATH="s3://pudl.catalyst.coop/$1/"
 
-    # If the old outputs don't exist, these will exit with status 1, so we
-    # don't && them with the rest of the commands.
-    echo "Removing old outputs from $GCS_PATH."
-    gsutil -m -u "$GCP_BILLING_PROJECT" rm -r "$GCS_PATH"
-    echo "Removing old outputs from $AWS_PATH."
-    aws s3 rm "$AWS_PATH" --recursive
-
-    echo "Copying outputs to $GCS_PATH:" && \
-    gsutil -m -u "$GCP_BILLING_PROJECT" cp -r "$PUDL_OUTPUT/*" "$GCS_PATH" && \
-    echo "Copying outputs to $AWS_PATH" && \
-    aws s3 cp "$PUDL_OUTPUT/" "$AWS_PATH" --recursive
+    # Only attempt to update outputs if we have an argument
+    # This avoids accidentally blowing away the whole bucket if it's not set.
+    if [[ -n "$1" ]]; then
+        # If the old outputs don't exist, these will exit with status 1, so we
+        # don't && them with the rest of the commands.
+        echo "Removing old outputs from $GCS_PATH."
+        gsutil -m -u "$GCP_BILLING_PROJECT" rm -r "$GCS_PATH"
+        echo "Removing old outputs from $AWS_PATH."
+        aws s3 rm --recursive "$AWS_PATH"
+
+        echo "Copying outputs to $GCS_PATH:" && \
+        gsutil -m -u "$GCP_BILLING_PROJECT" cp -r "$PUDL_OUTPUT/*" "$GCS_PATH" && \
+        echo "Copying outputs to $AWS_PATH" && \
+        aws s3 cp --recursive "$PUDL_OUTPUT/" "$AWS_PATH"
+    else
+        echo "No distribution path provided. Not updating outputs."
+        exit 1
+    fi
+}
+
+function distribute_parquet() {
+    PARQUET_BUCKET="gs://parquet.catalyst.coop"
+    # Only attempt to update outputs if we have a real value of BUILD_REF
+    # This avoids accidentally blowing away the whole bucket if it's not set.
+    echo "Copying outputs to parquet distribution bucket"
+    if [[ -n "$BUILD_REF" ]]; then
+        if [[ "$GITHUB_ACTION_TRIGGER" == "schedule" ]]; then
+            # If running nightly builds, copy outputs to the "nightly" bucket path
+            DIST_PATH="nightly"
+        else
+            # Otherwise we want to copy them to a directory named after the tag/ref
+            DIST_PATH="$BUILD_REF"
+        fi
+        echo "Copying outputs to $PARQUET_BUCKET/$DIST_PATH" && \
+        gsutil -m -u "$GCP_BILLING_PROJECT" cp -r "$PUDL_OUTPUT/parquet/*" "$PARQUET_BUCKET/$DIST_PATH"
+
+        # If running a tagged release, ALSO update the stable distribution bucket path:
+        if [[ "$GITHUB_ACTION_TRIGGER" == "push" && "$BUILD_REF" == v20* ]]; then
+            echo "Copying outputs to $PARQUET_BUCKET/stable" && \
+            gsutil -m -u "$GCP_BILLING_PROJECT" cp -r "$PUDL_OUTPUT/parquet/*" "$PARQUET_BUCKET/stable"
+        fi
+    fi
 }
```

Comment:
First I tried moving the wholesale removal of the `parquet/` directory from the output cleanup into the deployment function, but this didn't work because of the need to deploy twice when we have a versioned release (once to the versioned path, and once to `stable`)

Then I tried more selectively deploying outputs in the existing deployment function, instead of just copying everything in the directory, but this didn't work because [the S3 CLI (incredibly) can't understand wildcards](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/s3/index.html#use-of-exclude-and-include-filters) and there was no simple way to specify the set of parquet vs. non-parquet files for separate distribution.

Finally I made this terrible cut-and-paste copy of the deployment function which gets called before the output cleanup.

Is there a simple better way? Short of re-writing the entire thing in Python and bailing on these CLIs?

Reply:
I was referring to `copy_outputs_to_distribution_bucket()` which calls `upload_to_dist_path()` twice with different arguments if we're doing a versioned (aka `stable`) release.